Track Switching via the SWITCH_FROM parameter (#1674)
moq-wg/moq-transport: draft-ietf-moq-transport.md

the result of the request, some Objects could be received even if the FETCH
results in error.

## Track Switching {#track-switching}

A subscriber can atomically stop delivery on one subscription (the suspend

> subscription) and start delivery on another (the resume subscription) by

Reviewer: nit: resume is potentially confusing, because sometimes it is a new subscription.

Maybe 'join' or 'joining' or 'starting' subscription?

including the SWITCH_FROM parameter ({{switch-from}}) in a SUBSCRIBE or
REQUEST_UPDATE on the resume subscription's request stream.  This enables track
switching (e.g., ABR quality changes, alternate camera angles, or meeting
participants) without tearing down subscriptions or issuing standalone fetches.

On receiving a message containing SWITCH_FROM, the publisher:

1. Validates that Switch From Request ID identifies an existing subscription
   and is not the same as the resume subscription's Request ID. If not,
   responds with REQUEST_ERROR `INVALID_SWITCH`. The resume subscription

>    MUST be transitioning from Forward State 0 to 1 as part of the switch; if

Reviewer 2: I think you might want to rephrase this as "The resume subscription MUST either be a new subscription with a Forward State of 1 or be transitioning from Forward State 0 to 1 as part of the switch". The existing sentence applies for REQUEST_UPDATE, but not for new subscriptions.

>    MUST be transitioning from Forward State 0 to 1 as part of the switch; if

Reviewer: Do you mean MUST, or just that the SWITCH_FROM causes the forward state to change?

Also, from a design perspective I think it'd be ideal for the publisher to send a REQUEST_UPDATE with the new end of the filter.

Author: The subscription in SWITCH_FROM has to start in 0 and end in 1.  We could just say Forward is ignored if it accompanies a SWITCH_FROM, and the resume subscription is immediately paused if it's in Forward=1 when SWITCH_FROM is received.  Is that better?

> Also, from a design perspective I think it'd be ideal for the publisher to send a REQUEST_UPDATE with the new end of the filter.

The new _end_ of the _pause subscription's_ filter?  It's a little odd, since that's not really a request update, it's a "state update" (eg it does not solicit an OK or ERROR, or actually update anything).  If we really need that, we should add a new message (top-N can use this also when it toggles Forward?).

   it already has Forward State 1, or the message carrying SWITCH_FROM has

>    FORWARD = 0, the publisher responds with REQUEST_ERROR `INVALID_SWITCH`.

Reviewer 2: I think you might want to update the wording under the "Updating Subscriptions" section of the spec, which says: "When a REQUEST_UPDATE is unsuccessful, the publisher MUST also terminate the subscription by sending a PUBLISH_DONE with error code UPDATE_FAILED". This kind of contradicts that section.

Author: See Ian's point above, we could also remove this as an error condition.

> 2. Sets resume to Forward State 1 and applies the SUBSCRIPTION_FILTER. This

Reviewer: ```suggestion
2. Sets the resume Subscription to Forward State 1 and applies the SUBSCRIPTION_FILTER. This
```

   ensures objects after the Current Group are not missed.

3. Waits until it is ready to publish an object from the SUBSCRIPTION_FILTER's
   Start Location's Group (the Start Group; see {{subscription-filters}}),
   computed from the resume track at the time the request is received, while
   continuing to deliver objects on the suspend subscription.  When GROUP_ORDER
   is Descending and Start Group is before the Current Group, the publisher
   waits for the first object in the largest fill group instead.

4. Stops delivery on the suspend subscription:

   * Mode Hard (0x0): sets Forward State 0 on the suspend subscription. If

>      Publish Done is 1, the publisher also sends PUBLISH_DONE.

Reviewer 2: Which status code should be used in the PUBLISH_DONE message? How about we define a new status code SWITCHING_AWAY?

Reviewer: Does this change in Forward state get communicated via REQUEST_UPDATE?

Author: > Does this change in Forward state get communicated via REQUEST_UPDATE?

Not presently. See above.

   In all modes, the publisher also resets any outstanding suspend data for

>    groups greater than or equal to the Start Group; objects already in flight

Reviewer: This is the Start Group specified in the new subscription?

```suggestion
   In all modes, the publisher also resets any outstanding data streams for
   groups greater than or equal to the Start Group; objects already in flight
```

Author: > This is the Start Group specified in the new subscription?

The computed start group in the "resume" subscription, yes.

Author: Also I realized that this cancellation behavior may be incorrect for Hard switch -- we should cancel _all_ outstanding streams?

   can still be received by the subscriber.

   If Publish Done is 0, the suspend subscription remains established.

5. Begins delivery of resume subscription from Start Group. If the
   SUBSCRIPTION_FILTER is a fill filter type, opens a fill fetch stream using
   the resume subscription's Request ID (see {{fill-semantics}}), if necessary.

6. Responds with SUBSCRIBE_OK or REQUEST_OK as appropriate, including
   LARGEST_OBJECT if the SUBSCRIPTION_FILTER is a fill filter type.

> If the publisher times out waiting to be ready to publish an object from the

Reviewer: to be ready?

> If the publisher times out waiting to be ready to publish an object from the

Author: ```suggestion
If the publisher times out waiting to publish an object from the
```

the result of the request, some Objects could be received even if the FETCH
results in error.

## Track Switching {#track-switching}

> A subscriber can atomically stop delivery on one subscription (the suspending

Reviewer 3: ```suggestion
A subscriber can automically stop delivery on one subscription (the suspending
```

Author: automically is not a word?  atomic (as a single operation) is what I want.

> subscription) and start delivery on another (the activating subscription) by

Reviewer 4: nit: add italics for "suspending"/"activating" terms.

Actually, shouldn't it be "suspended" and not "suspending"?  "Suspending" implies it's the subscription that has caused suspension, not the one being suspended.

Author: Italics is a good call. At the beginning of the switch, the suspending stream is active and the activating stream is not, so I think "-ing" is correct.

> subscription) and start delivery on another (the activating subscription) by

Reviewer 3: we need terminology on what is suspending subscriptions

Author: > A subscriber can atomically stop delivery on one subscription (the suspending subscription) and start delivery on another (the activating subscription) 

I'm not sure what else you want me to say, I find it clear enough.

> subscription) and start delivery on another (the activating subscription) by

Author: ```suggestion
A subscriber can atomically stop delivery on one subscription (the _suspending
subscription_) and start delivery on another, possibly new subscription (the _activating subscription_) by
```

including the SWITCH_FROM parameter ({{switch-from}}) in a SUBSCRIBE or
REQUEST_UPDATE on the activating subscription's request stream.  This enables

> track switching (e.g., ABR quality changes, alternate camera angles, or meeting

Reviewer 3: This doesn'tt allow all forms of ABR .. may be we should call it Client Side ABR ?

> track switching (e.g., ABR quality changes, alternate camera angles, or meeting

Author: ```suggestion
subscriber track switching (e.g., ABR quality changes, alternate camera angles, or meetings
```

> participants) without tearing down subscriptions or issuing standalone fetches.

Reviewer 5: I think it would be best to define the IANA table for types of switches in PR and have just one entry in it of hard but that sets up the structure for later PRs.

> participants) without tearing down subscriptions or issuing standalone fetches.

Reviewer 3: IIUC, we tear down the subscription when PUBLISH_DONE is set to true, isn't it ?

Reviewer 3: also I might be missing the context with standalone fetches here. We should expand a bit more for reader to understand ( and yes not in full details :-) )

> participants) without tearing down subscriptions or issuing standalone fetches.

Author: ```suggestion
participants) without coordinating multiple messages.
```

> On receiving a message containing SWITCH_FROM, the publisher:

Reviewer 5: Are all publishers required to support this or just relays ?  I think it would be best to have it like it is that everything needs to support this but I can imaging lots of end user clients that are original publishers uploading video might not want to support this. Also - I am happy if we merge this PR without solving this and just raise this as an open issue that can be dealt with later.

Author: Discussed 7/6 - no one raised a significant objection to all publishers having to support this.

Reviewer 3: I do think, it will be helpful for a publisher to reject SWITHC_FROM even after saying yes initially. We need to define a way to do it. I can think of upstream goaway subscription doesn't has the quality to be switched_to anymore.

Author: > I do think, it will be helpful for a publisher to reject SWITHC_FROM even after saying yes initially.

There isn't really a "saying yes initially" -- the _final_ step in the algorithm is sending REQUEST_OK (or SUBSCRIBE_OK). If the switch becomes impossible before it completes, send REQUEST_ERROR. The quality being "switched to" is the track this parameter is operating on -- so ERROR/PUBLISH_DONE is a fine way to signal that.

> On receiving a message containing SWITCH_FROM, the publisher:

Reviewer 6: The publisher actions here are numbered 1 to 6, but the text does not say whether each step must wait for the previous one to complete before starting. For Hard mode this does not matter, but for Soft mode (#1675) the ordering is important: the suspending track must finish draining (step 4) before the activating track starts delivery (step 5).

Could we state whether these steps are strictly sequential? See the related comment on #1675: https://github.com/moq-wg/moq-transport/pull/1675#discussion_r3597470349

Author: I replied in 1675 ([comment](https://github.com/moq-wg/moq-transport/pull/1675#discussion_r3608721773)).

Without improved end-of-group signaling from publisher to relay, forcing these steps to be sequential will introduce implementation specific delays.

1. Validates that Switch From Request ID identifies an existing subscription
   and is not the same as the activating subscription's Request ID. If not,
   responds with REQUEST_ERROR `INVALID_SWITCH`. The activating subscription
   handles Forward State transitions as part of the switch; if the message

>    carrying SWITCH_FROM has a FORWARD parameter, the publisher responds with

Reviewer 6: "The activating subscription handles Forward State transitions as part of the switch" is vague — "handles" does not say what actually happens. Please state it directly.

I am also not sure FORWARD should be rejected at all. The suspending subscription is allowed to already be in Forward State 0, so why is Forward State 0 not allowed on the activating subscription? With `FORWARD=0` and Publish Done=1, the switch would stop the old track and leave a paused subscription on the new track — a silent switch from one track to another, without starting delivery yet. That seems like a valid case. Can we clarify why FORWARD is forbidden, and whether `FORWARD=0` on the activating subscription should be allowed?

Reviewer 3: Agree with @gwendalsimon  and not been able to convince myself on the reasoning. May be I am missing some context here.

Author: > The suspending subscription is allowed to already be in Forward State 0, so why is Forward State 0 not allowed on the activating subscription? 

The action of a SWITCH_FROM is not necessary if the activating subscription is going to forward = 0 -- there is nothing to coordinate, and logic about "when there's an object to send on the activating subscription" doesn't make any sense.  If folks strongly feel there is a use case we can allow it, but I think it will add a bunch of text for a corner case that was not discussed as a use case.

>    REQUEST_ERROR `INVALID_SWITCH`.  Note it is not an error if the pausing

Reviewer 3: s/pausing/suspending ?

>    REQUEST_ERROR `INVALID_SWITCH`.  Note it is not an error if the pausing

Author: ```suggestion
   REQUEST_ERROR `INVALID_SWITCH`.  Note it is not an error if the suspending
```

   subscription is in Forward State 0.

> 2. Sets the activating subscription to Forward State 1 and applies the

Reviewer 3: This is assuming we have an activating subscription already with forward = 0. A Switch is possible even for a new subscription.

An activating subscription can still have forward=0.

Author: > This is assuming we have an activating subscription already with forward = 0. A Switch is possible even for a new subscription.

If the SWITCH_FROM is accompanying a SUBSCRIBE, the same logic applies.  I added "possibly new subscription" to the opening paragraph to further establish that.

> An activating subscription can still have forward=0.

This PR does not allow that currently - SWITCH_FROM always sets the new subscription to Forward=1.

>    SUBSCRIPTION_FILTER. This ensures objects after the Current Group are not

Reviewer 5: at this point in the text, it's not clear to me what SUBSCRIPTION_FILTER we are talking about. I think we just need bit more clarification text, perhaps above this numbered list, to explain the filter part.

>    current SUBSCRIPTION_FILTER. This ensures objects after the Current Group

Reviewer 3: we need to clarify on what is "current" here ? Is it the latest object filter ?

Author: The "current" subscription filter is either the SUBSCRIPTION_FILTER param in the same message as SWITCH_FROM, or the most recently sent SUBSCRIPTION_FILTER.

>    current SUBSCRIPTION_FILTER. This ensures objects after the Current Group

Author: ```suggestion
   the most recent SUBSCRIPTION_FILTER parameter for this subscription (which
   can be in the same message carrying SWITCH_FROM). This ensures objects
   in Groups greater than or equal to the Largest Object's Group
```

>    missed.

Author: ```suggestion
2. Sets the activating subscription to Forward State 1 and applies the
   current SUBSCRIPTION_FILTER. This ensures objects after the Current Group
   are not missed.
```

>    are not missed.

Reviewer 4: Since this already depending on fill-fetch, maybe "applies the current filter" -> "creates fill-fetch streams as appopriate"?  Otherwise it sounds ambiguous as to what we're actually doing.

Reviewer 3: I also think, this doesn't handle the case on what happens to in-progress fill stream or when the new subscription cannot be switched because fill failed . It would be good to enumerate the edge/error cases.

Reviewer 6: I agree this is ambiguous. "current" is not defined, and "applies" does not say what happens in practice. If the intent is to create fill-fetch streams, it would be better to say it explicitly, as you suggest.

Separately, the sentence "This ensures objects after the Current Group are not missed" is hard to understand. First, what is the Current Group? It is written here as a defined term, but it is not defined in the draft. Second, "missed" is unclear — do we mean the objects are not sent to the subscriber?

>    are not missed.

Author: ```suggestion
   are not missed during the transition.
```

3. Waits until it is ready to publish an object from the SUBSCRIPTION_FILTER's
   Start Location's Group (the Start Group; see {{subscription-filters}}),
   computed from the activating track at the time the request is received, while
   continuing to deliver objects on the suspending subscription.  When GROUP_ORDER
   is Descending and Start Group is before the Current Group, the publisher

>    waits for the first object in the largest fill group instead.

Reviewer 5: Could we just remove the support for Descending here? Any use case that needs it ? Do it in later PR ? My issue is thinking about how implementations are going to buffer waiting for largetst fill group sounds painful. I don't think about Descending much so I may have this wrong but it does not seem like this case even makes any sense in hard mode. Can it be in a different mode.

Author: The nature of Descending is that it makes one's head hurt.  Anyone whose head is not hurting just hasn't tried to implement it yet.

I don't the the utility of Descending mode is different in Hard vs Soft switch, since it's about the activating subscription "Do I have the first requested object ready yet".  I don't think it's any more complicated than Descending support in general -- if you have a cache, then you likely have the object (it's from a more recent group) and if you don't you can issue a Descending FETCH upstream.  There shouldn't be a "buffer incoming objects" case?

Author: Discussed on 7/6 - Descending needs to be supported.

Reviewer 6: "the largest fill group" is not defined. The fill concepts this builds on (fill range, Largest Object, Group Order) are defined in #1673, but "fill group" is not defined there either.

In Descending order the highest group is delivered first (#1673), so I assume "largest fill group" means the highest group in the fill range, that is, the group of the Largest Object. It would be clearer to express this using the terms from #1673 rather than introduce a new term here.

Reviewer 3: I may be misunderstood here. I felt no one had opinion on it . Apologies if my recollection is wrong here. Unfortunately, I am not seeing the meeting minutes to confirm either.

Author: > I may be misunderstood here. I felt no one had opinion on it . Apologies if my recollection is wrong here. Unfortunately, I am not seeing the meeting minutes to confirm either.

I rewatched the video.  No one expressed a verbal opinion besides me.  There was text in Zulip stating "Descending is required to support live".  I said I was inclined to leave it in.  No one objected.

>    waits for the first object in the largest fill group instead.

Author: ```suggestion
   waits for the first object in the largest group in the fill range instead.
```

Updating to use 1763 "largest group in the fill range".

4. Stops delivery on the suspending subscription:

>    * Mode Hard (0x0): sets Forward State 0 on the suspending subscription. If

Reviewer 4: "Immediate"?  "Hard" seems like a vague term in absence of other modes.

Reviewer 6: I agree "Hard" is not clear. I am not sure "Immediate" is the best replacement either, because it describes *when* the switch happens, not what happens at the junction between the two tracks.

The defining property of this mode is that it does not align the two tracks. The suspending track is stopped as soon as the activating track is ready. Its last object can be before the activating track's first object (a gap), or after it (an overlap that the subscriber may need to drop from its buffer).

Because the essence is alignment, not timing, I would suggest "Unaligned". The second mode is already proposed in #1675 (Soft Mode); an "Unaligned" / "Aligned" pair would describe the two modes more clearly than "Hard" / "Soft".

Author: I like "Unaligned" and "Aligned". Hard is Unaligned because it doesn't require aligned group numbers between the tracks in order to be effective (though it can still work if they are).  Soft mode mostly makes sense when the tracks have aligned groups, since it uses a group number from one Track to set the end filter on another.

Reviewer 5: let me put a very weak vote for Immediate but don't really care.

>    * Mode Hard (0x0): sets Forward State 0 on the suspending subscription. If

Reviewer 3: We need to explain when the relay decides to it. Is it at the point of first object being delivered on the activating subscription to ensure data flow has started ? or is it at the point of receiving ack for bytes sent  on the beginning of the object.

Author: > We need to explain when the relay decides to it.

This is step 4 in an algorithm.  It happens as soon as Step 3 (`Waits until it is ready to publish an object ...`) is complete.

> or is it at the point of receiving ack for bytes sent on the beginning of the object.

There's nowhere else in MOQ where we require the timings of Acks so it's definitely not that.

     Publish Done is 1, the publisher also sends PUBLISH_DONE with code

>      SUBSCRIPTION_ENDED.

Reviewer 3: may be good to add a reason as SWITCHED ?

Author: I looked at this before and decided not to add a new code at this time, but now I don't remember exactly why.

>    In all modes, the publisher also resets any outstanding streams on the suspending

Author: ```suggestion
   In all modes, the publisher also resets any outstanding streams (including fill-fetch streams) on the suspending
```

>    subscription; objects already in flight can still be received by the subscriber.

Reviewer 3: isn't pausing or sending publish_done as implcit call to reset any ongoing streams ?
I wonder if it is safe for resetting the stream before the current object is completed ? I suspect there will be glitch or bad end user experience ? Perhaps this is a question to @gwendalsimon

Reviewer 3: we also need to tell what happens to fill streams too in this case.

Author: > isn't pausing or sending publish_done as implcit call to reset any ongoing streams ?

That text does not appear anywhere in the draft.  PUBLISH_DONE to the contrary has explicit text about waiting for open streams to complete so cancellation is clearly not for that case.

Pausing by itself does not trigger stream resets on open streams, but the publisher has to reset them if it gets another object (since skipping publishing it would create a gap for a paused subscriber).  An implementation is free to proactively reset on pause.  In any case, this is a separate issue.

>    If Publish Done is 0, the suspending subscription remains established.

Reviewer 6: The handling of Publish Done is split across two indentation levels, which makes step 4 hard to follow.

The "Publish Done is 1" case is nested under the Mode Hard bullet, so it reads as specific to Hard mode. The "Publish Done is 0" case is a separate paragraph at the step level. Between them there is an unrelated "In all modes" paragraph about resetting streams.

Publish Done is a property of the SWITCH_FROM parameter, not of the mode. It would be clearer to describe it once, at a single indentation level, independent of the mode, and keep the mode bullets for mode-specific behavior only (for Hard, setting Forward State 0).

Author: The multiple nesting levels are an artifact of splitting this PR.  The not-indented section applies to both Hard and Soft modes.

> It would be clearer to describe it once, at a single indentation level, independent of the mode, and keep the mode bullets for mode-specific behavior only 

That's sort of how it's written now?  The 0 behavior is common but the 1 behavior is not.  I'm inclined to leave this more-or-less as is, and deal with the flow after Soft also lands (or if it fails to reach consensus/moves to an extension).

Reviewer 6: Yes, let's see after Soft (unaligned) mode lands

5. Begins delivery of activating subscription from Start Group. If the
   SUBSCRIPTION_FILTER is a fill filter type, opens a fill fetch stream using

>    the resuming subscription's Request ID (see {{fill-semantics}}), if necessary.

Reviewer 6: `resuming` should be `activating`, to match the term used in the rest of this section.

```suggestion
   the activating subscription's Request ID (see {{fill-semantics}}).
```

>    the resuming subscription's Request ID (see {{fill-semantics}}), if necessary.

Reviewer 3: resuming subscription is a new term. I think we are referring to following 4 variants 
activating , suspending, pausing , resuming . Probably defining these at the beginning might be useful in general

Author: I switched from pausing/resuming to suspending/activating and missed a couple.  The terminology is introduced at the beginning of the section.

>    the resuming subscription's Request ID (see {{fill-semantics}}), if necessary.

Reviewer 3: Not sure I understand "if necessary" here. Why is it optional ? I suspect, if there is a fill parameter, the relay MUST setup fill fetch for the requested range, isn't it ?

> 6. Responds with SUBSCRIBE_OK or REQUEST_OK as appropriate, including

Reviewer 3: Responds with SUBSCRIBE_OK or REQUEST_OK to who ? activating subscription ? If so, its too much of time spent to respond I believe.

Author: Because the response carries LARGEST_OBJECT, it cannot be sent until that value is known.  There is a case where you could send the _OK before the fill once Largest Object is known, but in part the signal is delayed because it's communicating the time-point when the suspending subscription is paused.

If we add e.g. SUBSCRIPTION_STATE_UPDATE to communicate that event, we can return the _OK once Largest Object is determined.

> 6. Responds with SUBSCRIBE_OK or REQUEST_OK as appropriate, including

Author: ```suggestion
6. Responds on the activating subscription's control stream with SUBSCRIBE_OK or REQUEST_OK as appropriate, including
```

>    LARGEST_OBJECT if the SUBSCRIPTION_FILTER is a fill filter type.

Reviewer 4: I assume that for modes that do not immediately Forward=0 the old subscription, steps 5 and 6 would come before step 4.

Reviewer 3: +1 needs some clarification here

Author: > I assume that for modes that do not immediately Forward=0 the old subscription,

The Soft mode PR still has "4. Stops delivery on the suspending subscription" but goes on to detail that is accomplished by setting the end-filter.  So it's in the same immediacy as Hard (Forward=0).  The real questions are 1) do we want a separate control signal on the suspending subscription, and 2) can we then move step 6 earlier in either case?

If the publisher times out waiting to publish an object from the

> Start Group on the activating track, it MUST respond with REQUEST_ERROR `TIMEOUT`.

Reviewer 3: This is changing the behavior on how subscriptions are defined today. Today' we don't wait for object to eb ready to publish before responding to SUBSCRIBE. We wait for a response from upstream and object may or may not come anytime soon. 

I would like to learn more on why is this is mandated differently here.

Author: This is only in the context of SWITCH_FROM.  If you want uncoordinated suspend/activate - don't send SWITCH_FROM.

I mostly read your comment as wanting to get the _OK sooner, which I've responded to in the above comments.

> Start Group on the activating track, it MUST respond with REQUEST_ERROR `TIMEOUT`.

Reviewer 3: may be we already have this . What timeout is this ? randevouz timeout ?

Author: RENDEZVOUS_TIMEOUT is how long the subscriber is willing to wait at a relay with no known publishers for a track.

This is different -- it's that the activating track didn't receive the first object in their requested range in a timely way, and hence the old track was never suspended.  The publisher may be known, but the fill-fetch took too long, or e.g. NextGroup was used and the publisher didn't mint a new group in a reasonable time.

### Relay Handling of SWITCH_FROM {#relay-switch-from}

> When a relay receives a message containing SWITCH_FROM, it MUST NOT forward the

Reviewer 5: I think we need a SHOULD NOT instead of MUST NOT. We do the same thing for other things where a local wifi relay that is one the downstream side of a congested WAN link probably would want to forward this upstream.

> When a relay receives a message containing SWITCH_FROM, it MUST NOT forward the

Author: ```suggestion
When a relay receives a message containing SWITCH_FROM, it SHOULD NOT forward the
```

> SWITCH_FROM parameter upstream. The relay handles the switch locally, applying

Reviewer 3: SHOULD_NOT and relay handles locally are contradicting.

the start group computation from {{track-switching}} using its locally observed
state for the activating track and servicing any fill fetch stream from its cache

> and upstream sources as described in {{fill-semantics}}.

Author: ```suggestion
Relays ordinarily handle the switch locally, applying
the start group computation from {{track-switching}} using locally observed
state for the activating track and servicing any fill fetch stream from cache
and upstream sources as described in {{fill-semantics}}.  When a relay performs
the switch operation, it MUST NOT forward the SWITCH_FROM parameter upstream.
```

   and is not the same as the activating subscription's Request ID. If not,
   responds with REQUEST_ERROR `INVALID_SWITCH`. The activating subscription
   handles Forward State transitions as part of the switch; if the message
   carrying SWITCH_FROM has a FORWARD parameter, the publisher responds with
   REQUEST_ERROR `INVALID_SWITCH`.  Note it is not an error if the suspending
   subscription is in Forward State 0.

2. Sets the activating subscription to Forward State 1 and applies the
   the most recent SUBSCRIPTION_FILTER parameter for this subscription (which
   can be in the same message carrying SWITCH_FROM). This ensures objects
   in Groups greater than or equal to the Largest Object's Group
   are not missed during the transition.

3. Waits until it is ready to publish an object from the SUBSCRIPTION_FILTER's
   Start Location's Group (the Start Group; see {{subscription-filters}}),
   computed from the activating track at the time the request is received, while
   continuing to deliver objects on the suspending subscription.  When GROUP_ORDER
   is Descending and Start Group is before the Current Group, the publisher

>    waits for the first object in the largest group in the fill range instead.

Reviewer 7: Maybe it's easier to switch on any object in the fill range?

Reviewer 8: In the case where the payload is media and keyframes are located at group boundaries, wouldn't switching on 'any object' give you an un-decodeable object sequence? Since groups are defined agnostically as join points to tracks, shouldn't all switches occur at group boundaries?

4. Stops delivery on the suspending subscription:

   * Mode Hard (0x0): sets Forward State 0 on the suspending subscription. If
     Publish Done is 1, the publisher also sends PUBLISH_DONE with code
     SUBSCRIPTION_ENDED.

   In all modes, the publisher also resets any outstanding streams (including fill-fetch streams) on the suspending
   subscription; objects already in flight can still be received by the subscriber.

   If Publish Done is 0, the suspending subscription remains established.

5. Begins delivery of activating subscription from Start Group. If the
   SUBSCRIPTION_FILTER is a fill filter type, opens a fill fetch stream using
   the activating subscription's Request ID (see {{fill-semantics}}).

6. Responds on the activating subscription's control stream with SUBSCRIBE_OK or REQUEST_OK as appropriate, including
   LARGEST_OBJECT if the SUBSCRIPTION_FILTER is a fill filter type.

If the publisher times out waiting to publish an object from the

> Start Group on the activating track, it MUST respond with REQUEST_ERROR `TIMEOUT`.

Reviewer 5: This does not seem part of this PR. I would asume this happens the same as it would if forwarding is changes from 0 to 1.

### Relay Handling of SWITCH_FROM {#relay-switch-from}

Relays ordinarily handle the switch locally, applying
the start group computation from {{track-switching}} using locally observed

> state for the activating track and servicing any fill fetch stream from cache

Reviewer 5: This needs clear advice how it works for relays with no cache.

and upstream sources as described in {{fill-semantics}}.  When a relay performs

> the switch operation, it MUST NOT forward the SWITCH_FROM parameter upstream.

Reviewer 5: So the stub relays will need to forward this upstream. I think this MUST NOT needs to be a SHOULD NOT with an explanation around that

If omitted from SUBSCRIBE, the publisher's preference from
the Track is used. If omitted from FETCH, the receiver uses Ascending (0x1).

### SWITCH_FROM Parameter {#switch-from}

The SWITCH_FROM parameter (Parameter Type 0x24) consists of a varint Switch
From Request ID, a varint Mode, and a single byte of flags. It MAY appear in a
SUBSCRIBE or REQUEST_UPDATE (for a subscription) message.

~~~
SWITCH_FROM {
  Switch From Request ID (vi64),
  Mode (vi64),
  Publish Done (1),

>   Reserved (7),

Reviewer: Should we call it 'Reserved bits' or 'Reserved flags' to be clearer?

>   Reserved (7),

Author: RFC 9000 uses Reserved Bits

```suggestion
  Reserved Bits (7),
```

}
~~~

* Switch From Request ID: The Request ID of the subscription to suspend.

* Mode: A vi64 enum selecting how the suspend subscription is stopped. The
  following mode is defined: Hard (0x0). An endpoint that receives a Mode value

>   that is not defined MUST close the session with `PROTOCOL_VIOLATION`. See

Reviewer: ```suggestion
  that is not defined or negotiated MUST close the session with `PROTOCOL_VIOLATION`. See
```

Author: Isn't "or negotiated" implied by 

> Extensions can define new Message types, new Parameters, new Properties, or new framing for Streams and Datagrams.

Or maybe that sentence needs "Enum values"?

Reviewer: It is, but somewhere we should make it clear this can be extended.

Author: See #1799

If omitted from SUBSCRIBE, the publisher's preference from
the Track is used. If omitted from FETCH, the receiver uses Ascending (0x1).

### SWITCH_FROM Parameter {#switch-from}

> The SWITCH_FROM parameter (Parameter Type 0x24) consists of a varint Switch

Reviewer 9: Shouldn't the parameter type be an odd value?

Author: Parameters no longer use the even/odd distinction for varint/length prefixed.  Each parameter has a bespoke format that must be known in order to parse it.

Reviewer 9: I missed that change, my bad. I was still thinking of message parameters as KVPs.

From Request ID, a varint Mode, and a single byte of flags. It MAY appear in a
SUBSCRIBE or REQUEST_UPDATE (for a subscription) message.

~~~
SWITCH_FROM {
  Switch From Request ID (vi64),
  Mode (vi64),
  Publish Done (1),
  Reserved Bits (7),
}
~~~

* Switch From Request ID: The Request ID of the subscription to suspend.

* Mode: A vi64 enum selecting how the suspending subscription is stopped. Modes
  are registered in the IANA table "MOQT SWITCH_FROM Modes"
  ({{iana-switch-from-modes}}); the following mode is defined: Hard (0x0). An
  endpoint that receives a Mode value that is not defined MUST close the session
  with `PROTOCOL_VIOLATION`. See {{track-switching}}.

* Publish Done: If 1, the publisher sends PUBLISH_DONE on the suspend
  subscription as described in {{track-switching}}.

* Reserved Bits: MUST be 0. An endpoint that receives a non-zero value MUST

>   close the session with `PROTOCOL_VIOLATION`.

Reviewer 3: may be we can remove this field as we are defining anything .If needed, we can add this back as it is in parameter.

Author: It's needed for byte alignment

Reviewer 6: Can we have a SHOULD instead of a MUST for "the reserved bits at 0"? The reason is that it could be used for implementation-specific flags, which the Relay could use to run some optimized track transition (e.g. temporary priority management to accelerate the switch)

Author: @gwendalsimon if someone wants to use these bits, they need to negotiate an extension first.  This is consistent with the rest of the document, where we tend to allocate a full varint for an enum with 2 or 3 fields, and say any other value is a protocol violation.  It's fine to define extensions, you just need to ensure the peer speaks it first.

: In response to a FETCH, a relay publisher detected the track was
malformed (see {{malformed-tracks}}).

INVALID_SWITCH:
: In response to a SUBSCRIBE or REQUEST_UPDATE carrying SWITCH_FROM, the track
switch cannot be performed: the Switch From Request ID does not identify an
existing subscription, is the same as the resume subscription's Request ID, the

> resume subscription is not transitioning from Forward State 0 to 1, or the

Reviewer: These last two cases seem identical to me?

Author: They are not exactly the same.  But maybe it's moot if we decide to just ignore Forward?

: In response to a FETCH, a relay publisher detected the track was
malformed (see {{malformed-tracks}}).

INVALID_SWITCH:
: In response to a SUBSCRIBE or REQUEST_UPDATE carrying SWITCH_FROM, the track
switch cannot be performed: the Switch From Request ID does not identify an
existing subscription, is the same as the activating subscription's Request ID, or the

> request contained the FORWARD parameter (see {{track-switching}}).

Reviewer 3: nit: grammar seems correction here

> request contained the FORWARD parameter (see {{track-switching}}).

Author: ```suggestion
: In response to a SUBSCRIBE or REQUEST_UPDATE carrying SWITCH_FROM, the track
switch cannot be performed (see {{track-switching}}).
```

: In response to a FETCH, a relay publisher detected the track was
malformed (see {{malformed-tracks}}).

INVALID_SWITCH:
: In response to a SUBSCRIBE or REQUEST_UPDATE carrying SWITCH_FROM, the track

> switch cannot be performed (see {{track-switching}}).

Reviewer 5: how do we get the extra information of why the switch could not be done
